### PR TITLE
AO3-5056 External author can't update preferences while doing nothing with imported works

### PR DESCRIPTION
**Pull request by sarken**
## Issue

https://otwarchive.atlassian.net/browse/AO3-5056

## Purpose

* Removes the redirect that was preventing us from getting to the external author preference-updating code, causing the bug in the description
* Stops using `+=` on the flash error for updating external author preferences, because there was no empty `flash[:error] = ""` to `+` it to (and that's why the pending test was commented out)
* Adds a metric ton of tests

## Testing

Refer to JIRA

### CODE REVIEW

File: app/controllers/external_authors_controller.rb [hunk 1]
```diff
@@ -74,7 +73,10 @@ def update
       @external_author.delete_works
       flash[:notice] += "Your imported stories have been deleted. "
     end
-    @invitation.mark_as_redeemed if @invitation && !params[:imported_stories].blank?
+
+    if @invitation && params[:imported_stories].present? && params[:imported_stories] != "nothing"
```
Comment (sarken) [inline app/controllers/external_authors_controller.rb]: Justification of this style change: it was too gosh darn long to keep in the one-line format.

File: spec/controllers/external_authors_controller_spec.rb [hunk 2]
```diff
@@ -91,48 +91,138 @@
...
+          expect(response).to render_template :edit
+          expect(flash[:error]).to eq "There were problems saving your preferences."
+        end
+      end
     end
 
     context "when the user has permission through an invitation" do
       context "when doing nothing with imported works" do
-        it "redirects with a success message" do
+        it "does not mark invtiation redeemed and redirects with a success message" do
```
Comment (ariana-paris) [inline spec/controllers/external_authors_controller_spec.rb]: Typo in `invitation`

**Comment by sarken** [comment]
Wow, I have zero memory of making this pull request, but I will fix the typo and push!

**Comment by sarken** [comment]
Oh, sorry, I didn't see this was marked Approved with the review tool -- going to switch the label from "Coder has actioned review" to "Reviewed: Ready to Merge" since the action was a typo fix.

**Comment by sarken** [comment]
Closing so I can get under 5 pull requests and open a new one for an issue that's blocking two of my existing PRs.

**Comment by sarken** [comment]
This had merge conflicts which I've attempted to resolve, so I've flipped it away from Ready to Merge.

File: app/controllers/external_authors_controller.rb [hunk 3]
```diff
@@ -77,9 +74,12 @@ def update
...
       flash[:notice] += "Your imported stories have been deleted. "
     end
-    @invitation.mark_as_redeemed if @invitation && !params[:imported_stories].blank?
 
-    if @external_author.update_attributes(external_author_params[:external_author] || {})
+    if @invitation && params[:imported_stories].present? && params[:imported_stories] != "nothing"
+      @invitation.mark_as_redeemed
+    end
+
+    if @external_author.update_attributes(external_author_params[:external_author])
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply app/controllers/external_authors_controller.rb:82]: We should make this change (we'll need to by Rails 6).

File: features/importing/archivist.feature [hunk 4]
```diff
@@ -230,13 +230,69 @@ Feature: Archivist bulk imports
...
+      And I should see "Your imported stories have been deleted. Your preferences have been saved."
     When the email queue is clear
       And I am logged in as "archivist"
       And I import the work "http://ao3testing.dreamwidth.org/325.html" by "randomtestname" with email "random@example.com"
       And the system processes jobs
      Then 0 emails should be delivered to "random@example.com"
 
+  Scenario: Leave an imported work in the archivist's care
+    When I import the work "http://ao3testing.dreamwidth.org/593.html" by "randomtestname" with email "random@example.com"
+      And the system processes jobs
```
Comment (redsummernight) [inline features/importing/archivist.feature:243]: This step doesn't do anything anymore, we shouldn't use it ([AO3-5722](https://otwarchive.atlassian.net/browse/AO3-5722)).

File: spec/controllers/external_authors_controller_spec.rb [hunk 5]
```diff
@@ -71,91 +71,299 @@
...
+              imported_stories: "nothing",
+              external_author: {
+                do_not_email: 1,
+                do_not_import: 1
+              }
+            }
+
+            allow_any_instance_of(ExternalAuthor).to receive(:update_attributes).and_return(false)
+            put :update, params: parameters
+            allow_any_instance_of(ExternalAuthor).to receive(:update_attributes).and_call_original
```
Comment (redsummernight) [inline spec/controllers/external_authors_controller_spec.rb:225]: You can leave out the reset calls - stubs get undone at the end of a test anyway.

File: app/controllers/external_authors_controller.rb (hunk 1 above)
Comment (redsummernight) [reply app/controllers/external_authors_controller.rb:78]: FYI you can break the conditional into multiple lines as well, like:

https://github.com/otwcode/otwarchive/blob/bedbbef8c43585d97b5a56062355921e4c8f5fa7/app/models/creatorship.rb#L147-L149

**Review by redsummernight — APPROVED** [review]
Description:
> * Stops using `+=` on the flash error for updating external author preferences, because there was no empty `flash[:error] = ""` to `+` it to (and that's why the pending test was commented out)

This change was in #2754, so not in this PR anymore.

**Pull request merged**